File: Data Structures & Algorithms/unionFind/submission-5.java
class UnionFind {
    private Map<Integer, Integer> parent;
    private Map<Integer, Integer> rank;
    private int numOfComponents;

    public UnionFind(int n) {
        this.parent = new HashMap<>();
        this.rank = new HashMap<>();
        this.numOfComponents = n;

        for (int i = 0; i < n; i++) {
            this.parent.put(i, i);
            this.rank.put(i, 0);
        }
    }

    public int find(int x) {
        int p = this.parent.get(x);
        while (p != this.parent.get(p)) {
            // replaced by grand parent
            this.parent.put(p, this.parent.get(this.parent.get(p)));
            p = this.parent.get(p);
        }
        return p;
    }

    public boolean isSameComponent(int x, int y) {
        return this.find(x) == this.find(y);
    }

    public boolean union(int x, int y) {
        int p1 = this.find(x);
        int p2 = this.find(y);

        if (p1 == p2) return false;

        int r1 = this.rank.get(p1);
        int r2 = this.rank.get(p2);
        if (r1 > r2) {
            this.parent.put(p2, p1);
        } else if (r2 > r1) {
            this.parent.put(p1, p2);
        } else {
            this.parent.put(p2, p1);
            this.rank.put(p1, this.rank.get(p1) + 1);
        }
        this.numOfComponents--;
        return true;
    }

    public int getNumComponents() {
        return numOfComponents;
    }
}

File: Data Structures & Algorithms/greatest-common-divisor-traversal/submission-5.py
class Solution:
    def canTraverseAllPairs(self, nums: List[int]) -> bool:
        num_set = set(nums)
        num_set_list = list(num_set)
        n = len(num_set_list)
        if 1 in num_set:
            return len(nums) == 1
        uf = UnionFind(num_set_list)
        num_of_group = n

        def is_gcd_greater_than_1(a: int, b: int) -> bool:
            low, high = min(a, b), max(a, b)
            if low == 1 or high == 1:
                return False
            if high % low == 0:
                return True
            divisors = []
            for i in range(2, int(math.sqrt(low) + 1)):
                if low % i == 0:
                    divisors.append(i)
                    divisors.append(low // i)
            for d in divisors:
                if high % d == 0:
                    return True
            return False

        for i in range(n - 1):
            for j in range(i + 1, n):
                a, b = num_set_list[i], num_set_list[j]
                if is_gcd_greater_than_1(a, b) and uf.union(a, b):
                    num_of_group -= 1

        return num_of_group == 1

class UnionFind:
    
    def __init__(self, nums):
        self.parent = {}
        self.rank = {}
        for i in nums:
            self.parent[i] = i
            self.rank[i] = 1

    def find(self, x):
        p = self.parent[x]
        while p != self.parent[p]:
            self.parent[p] = self.parent[self.parent[p]]
            p = self.parent[p]
        return p

    def union(self, a, b):
        p1, p2 = self.find(a), self.find(b)
        if p1 == p2:
            return False
        r1, r2 = self.rank[p1], self.rank[p2]
        if r1 >= r2:
            self.parent[p2] = p1
            self.rank[p1] = r1 + r2
        elif r2 > r1:
            self.parent[p1] = p2
            self.rank[p2] = r1 + r2
        return True

        
